go/fn/run.go

```diff
@@ -42,5 +46,9 @@ func Run(p ResourceListProcessor, input []byte) ([]byte, error) {
 			rl.Results = append(rl.Results, ErrorResult(err))
 		}
 	}
-	return rl.ToYAML()
+	out, yamlErr := rl.ToYAML()
```

Comment:
This is same as what I did in https://github.com/kubernetes-sigs/kustomize/pull/4319/files#diff-ce9f57feb2483d04f1042f75772dd11b8b785de449a0a223d7297633f5f74a00R362-R369

Reply:
Take this comment with grain of salt (as I have limited understanding of the code at this layer), but seems like `p.Process(rl)` should return (result, err) ? overloading the returned err for error and result seems to make this code a bit complex.

Also noticed the new code in `kyaml` uses switch-case (looked easy to follow). May be worth updating ? 

/cc @yuwenma

Reply:
@droot I like your suggesting of separating the returned result and error. I left a comment https://github.com/GoogleContainerTools/kpt-functions-catalog/pull/782#discussion_r832646437

Reply:
@yuwenma @mengqiy I updated this PR in line with @droot's suggestion to have `Process` return both a `Result` and an `error` to simplify the `AsMain` code. PTAL. 

We can sync offline if needed.

Reply:
> Process(rl *ResourceList) (*Result, error)

This interface can't handle multiple Results. We have functions like gatekeeper that may want to return a list of results.

There are a few questions we should answer before we make a decision:
- If the developer return a Result here, does he/she also need to add it in the ResourceList?
- If Result has severity "error", but the returned error is nil, what does it mean?
- If Result has non-error severity, but the returned error is non nil, what should we do?
- Similar questions if we want to support returning a list of Results.

IMO if having some complexity in our code can the KRM function developers' life easier, it worths it.

Another option is using an interface like `Process(rl *ResourceList) (bool, error)`.
- Boolean can be an easier way to indicate if there is a failure.
- The error can be used by user to return arbitrary error. User should not return structured results here.
- Users are responsible for adding the structured results in the ResourceList.

Reply:
@yuwenma WDYT?

Reply:
Sorry for the late reply.

Can't agree more on the rationale why not return both Result and error. Duplicate user patterns introduces many caveats and requires additional risk condition handling. Mengqi has given some examples.
   
The if/else results approach is very handy for users and I don't think we want to remove that.

I'm actually thinking of changing the `Run` and `AsMain` workflow as following
1.  `Run` only returns `[]byte` even if it has errors. 
2. For critical failure inside `Run`, use panic/recovery like https://github.com/GoogleContainerTools/kpt-functions-sdk/pull/531
3. Provide a new go function to decide whether a resourcelist contains Error severity results
4. `AsMain` does not handle errors from `Run`

Reply:
@yuwenma Question, maybe I am misunderstanding your suggestion, but if `Run` does not return an error, how can users use `AsMain`? The function needs to return a nonzero exit code when there is an error according to the function spec. 

Current pattern is to do something like 

```
if err := AsMain(....); err != nil {
   os.Exit(1)
}
```

If `AsMain` does not have access to the errors from `Run`, then the error from `Run` will be swallowed, so how can function authors check for function error when deciding what exit code to return?

Reply:
That case falls in #2 "For critical failure inside Run, use panic/recovery like https://github.com/GoogleContainerTools/kpt-functions-sdk/pull/531"

Reply:
So if a `GetOrDie` panics in `p.Process(rl)`, `Run` will return an error and this triggers the `kpt fn eval` '[FAIL]` stdout.

So we basically use panic/recover to distinguish between panic and `rl.ToYAML` error. The code is like
```go
func Run(p ResourceListProcessor, input []byte) ([]byte, error) {
     ...
     defer handleOptOrDieErr()
	 err = p.Process(rl)
     defer handleCriticalErr()
     p.LogFnResults(err)
     return rl.ToYAML()
}
```

Reply:
> That case falls in #2 "For critical failure inside Run, use panic/recovery like #531"

Oh, I see. That makes sense. 

Currently `AsMain` writes the ResourceList to stdout, so do we want to move that to `Run`? We will need to write it to stdout before exiting so kpt can get the Results.

Reply:
E.g. 
```
func Run(p ResourceListProcessor, input []byte) ([]byte, error) {
     ...
     defer handleOptOrDieErr()
	 err = p.Process(rl)
     defer handleCriticalErr()
     p.LogFnResults(err)
     out, err := rl.ToYAML()
     _, err = os.Stdout.Write(out)
    return err
}
```